File: 966-binary-subarrays-with-sum/binary-subarrays-with-sum.java
class Solution {
    public int numSubarraysWithSum(int[] nums, int goal) {
        return count(nums,goal)-count(nums,goal-1);
    }
    public int count(int[] nums,int goal){
        int end=0;
        int start=0;
        int ans=0;
        int sum=0;
        while(end<nums.length){
            sum+=nums[end];
            while(start<=end && sum>goal){
                sum-=nums[start];
                start++;
            }
            ans+=end-start+1;
            end++;
        }
        return ans;
    }
}

File: 966-binary-subarrays-with-sum/binary-subarrays-with-sum.cpp
class Solution {
public:
    int numSubarraysWithSum(vector<int>& nums, int goal) {
        unordered_map<int,int> prefix;
        int c=0;
        int sum=0;
        prefix[0]=1;
        for(int i=0;i<nums.size();i++){
            sum+=nums[i];
            if(prefix.count(sum-goal)){
                c+=prefix[sum-goal];
            }
            prefix[sum]++;
        }
        return c;
    }
};